connect/transforms/src/main/java/org/apache/kafka/connect/transforms/TimestampConverter.java

```diff
@@ -326,38 +331,47 @@ protected R newRecord(R record, Schema updatedSchema, Object updatedValue) {
     protected abstract R newRecord(R record, Schema updatedSchema, Object updatedValue);
 
     private R applyWithSchema(R record) {
-        final Schema schema = operatingSchema(record);
+        final Schema originalSchema = operatingSchema(record);
+        final Object rawValue = operatingValue(record);
+        // Entire value is timestamp
         if (config.field.isEmpty()) {
-            Object value = operatingValue(record);
             // New schema is determined by the requested target timestamp type
-            Schema updatedSchema = TRANSLATORS.get(config.type).typeSchema();
-            return newRecord(record, updatedSchema, convertTimestamp(value, timestampTypeFromSchema(schema)));
-        } else {
-            final Struct value = requireStruct(operatingValue(record), PURPOSE);
-            Schema updatedSchema = schemaUpdateCache.get(value.schema());
-            if (updatedSchema == null) {
-                SchemaBuilder builder = SchemaUtil.copySchemaBasics(schema, SchemaBuilder.struct());
-                for (Field field : schema.fields()) {
-                    if (field.name().equals(config.field)) {
-                        builder.field(field.name(), TRANSLATORS.get(config.type).typeSchema());
-                    } else {
-                        builder.field(field.name(), field.schema());
-                    }
-                }
-                if (schema.isOptional())
-                    builder.optional();
-                if (schema.defaultValue() != null) {
-                    Struct updatedDefaultValue = applyValueWithSchema((Struct) schema.defaultValue(), builder);
-                    builder.defaultValue(updatedDefaultValue);
-                }
+            Schema updatedSchema = TRANSLATORS.get(config.type).typeSchema(originalSchema.isOptional());
+            return newRecord(record, updatedSchema, convertTimestamp(rawValue, timestampTypeFromSchema(originalSchema)));
+        }
+        // Value is Struct, only its single field should be converted
+        if (rawValue == null) {
+            Schema updatedSchema = updateSchema(originalSchema);
```

Comment:
This is not using a cached schema, which means that a new schema has to be built for every record that uses some original schema, A, but that also has a null timestamp field. Wouldn't it make sense to cache this updated schema? And, since the updates schema is not a function of the record, we should be able to look for the updated schema in the cache or update the schema and cache it before dealing with the raw value.

This might also mean that there are fewer lines changed, especially if we keep the `else` block and previous indentation (as mentioned before, even though strictly speaking the `else` is unnecessary), which might help us ensure that the behavior doesn't change for the non-null timestamps.